### PR TITLE
do not AnnounceBlocks on null behaviors, Fix race condition expectation in MiningNodeWithOneConnectionAlwaysSynced, restore coinbase maturity after altering it in test

### PR DESCRIPTION
**Pull request by monsieurleberre**
restore coinbaseMaturity upon leaving the test

image

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs [hunk 1]
```diff
@@ -226,7 +226,8 @@ protected override void OnNextCore(Block block)
             }
 
             // Announce the blocks to each of the peers.
-            IEnumerable<BlockStoreBehavior> behaviours = peers.Select(s => s.Behavior<BlockStoreBehavior>());
+            IEnumerable<BlockStoreBehavior> behaviours = peers.Select(s => s.Behavior<BlockStoreBehavior>())
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs]: Can the behavior be null here?

Comment (noescape00) [reply src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs]: Those null behaviors- are they added from the peer discovery? If so- don't add connectionManagerBehavior during peer creation in discovery

Comment (monsieurleberre) [reply src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs]: OK I'll check in more details, the issue I was seeing was that upon Disposing of the node, there was still some batches left to send, and then this causes an exception as the behavior is being disposed so it ends up in a strange state.
I think this change is correct anyway, but we could do try and avoid these situation passing and handling correctly a cancellation token...
I'll check in more details what you were advising, thanks

File: src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs [hunk 2]
```diff
@@ -226,7 +226,8 @@ protected override void OnNextCore(Block block)
...
+                .Where(b => b != null).ToList();
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs]: ToList is not needed here

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletAddressGenerationAndFundsVisibility_Steps.cs [hunk 3]
```diff
@@ -29,6 +29,7 @@ public partial class Wallet_address_generation_and_funds_visibility : BddSpecifi
         private CoreNode sendingStratisBitcoinNode;
         private CoreNode receivingStratisBitcoinNode;
         private long walletBalance;
+        private long previousCoinBaseMaturity;
```
Comment (noescape00) [inline src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletAddressGenerationAndFundsVisibility_Steps.cs]: Not sure if changes in this file belong to this pr. Why are they here?

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs [hunk 4]
```diff
@@ -801,6 +801,9 @@ public void Disconnect(string reason, Exception exception = null)
 
             if (this.IsConnected) this.SetStateAsync(NetworkPeerState.Disconnecting).GetAwaiter().GetResult();
 
+            this.Behaviors?.ToList().ForEach(b => b?.Detach());
```
Comment (noescape00) [inline src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs]: Why? No need to detach here.

File: src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs [hunk 5]
```diff
@@ -226,9 +226,10 @@ protected override void OnNextCore(Block block)
             }
 
             // Announce the blocks to each of the peers.
-            IEnumerable<BlockStoreBehavior> behaviours = peers.Select(s => s.Behavior<BlockStoreBehavior>());
+            List<BlockStoreBehavior> behaviours = peers.Select(s => s.Behavior<BlockStoreBehavior>())
+                .Where(b => b != null).ToList();
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs]: ToList is not needed here

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs (hunk 4 above)
Comment (rowandh) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs]: `ToList()` should also use null propagation

Comment (rowandh) [inline src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs]: `ToList()` should also use null propagation

Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs]: +1 to @monsieurleberre for the courage, but -2 for the Linq usage and most importantly -99 for corrupting the code with this change.

File: src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs (hunk 5 above)
Comment (monsieurleberre) [reply src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs]: To List here is to prevent multiple enumeration : one with "Count", and one with foreach :)

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs (hunk 4 above)
Comment (monsieurleberre) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs]: +110 to boost my damaged ego, that was not my idea, and it is getting reverted ;)

Comment (monsieurleberre) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs]: true! I'll remove that line anyway, thanks

File: src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs [hunk 6]
```diff
@@ -226,9 +226,10 @@ protected override void OnNextCore(Block block)
             }
 
             // Announce the blocks to each of the peers.
-            IEnumerable<BlockStoreBehavior> behaviours = peers.Select(s => s.Behavior<BlockStoreBehavior>());
+            List<BlockStoreBehavior> behaviours = peers.Select(s => s.Behavior<BlockStoreBehavior>())
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs]: Select will throw if peers collection is modified during the foreach which is used in the select
https://github.com/Microsoft/referencesource/blob/master/System.Core/System/Linq/Enumerable.cs

You can rewrite this to avoid using linq and avoid having a problem like mentioned above

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletAddressGenerationAndFundsVisibility_Steps.cs [hunk 7]
```diff
@@ -38,6 +39,9 @@ protected override void BeforeTest()
 
         protected override void AfterTest()
         {
+            this.sendingStratisBitcoinNode.FullNode.Network.Consensus.CoinbaseMaturity = this.previousCoinBaseMaturity;
```
Comment (noescape00) [inline src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletAddressGenerationAndFundsVisibility_Steps.cs]: Changes in WalletAddressGenerationAndFundsVisibility_Steps.cs do not belong to this PR

File: src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs (hunk 6 above)
Comment (monsieurleberre) [reply src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs]: I think it is OK: Linq uses the GetEnumerator() function of the Enumerable, which is for ConcurrentHashSet the following
image

**Comment by monsieurleberre** [comment]
image

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletAddressGenerationAndFundsVisibility_Steps.cs (hunk 7 above)
Comment (monsieurleberre) [reply src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletAddressGenerationAndFundsVisibility_Steps.cs]: OK, you are right, however I am just trying to make easy to understand PRs and not waste too much time in the process. This is an important fix too because not restoring CoinBaseMaturity leads to the value being changed for all subsequent tests, and test failures. It is easy to understand this very local change...

I should rename the PR to the branch it is based on : integration-test-small-fix :)

Comment (noescape00) [reply src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletAddressGenerationAndFundsVisibility_Steps.cs]: Ah, so that's fixing integration tests that are currently failing on master?

File: src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs (hunk 6 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs]: Ah, yes, peer collection uses ConcurrentHashSet inside. Then ok, it should be good.

**Comment by noescape00** [comment]
I'm fine with the PR, please sync chains and run the tests. 

I'll approve as soon as I do same as well.

**Comment by monsieurleberre** [comment]
OK as far as I am concerned, it looks like it is working, let mw know what you see on your side, or just merge :) Thanks !

**Comment by noescape00** [comment]
Chains synced, failed to pass integration tests. Do they pass for you?

**Comment by monsieurleberre** [comment]
Yes they do, only the 4 in the screenshot don't pass...
